### PR TITLE
fix: nonce handling

### LINKED ISSUE #667: fix: nonces
Fix nonce error handling. Batcher never recovers from reverted tx. Proposed solution is this:

image

### PR DESCRIPTION
**Pull request by taturosati**
# Implementation

Implement nonce cache clearing on fail, both on client and on batcher

image

# To test

Test normal flow, make some txs revert by sending same burst twice

### CODE REVIEW

File: batcher/aligned-batcher/src/lib.rs [hunk 1]
```diff
@@ -496,6 +496,20 @@ impl Batcher {
         Ok(())
     }
 
+    async fn reset_state(&self) {
```
Comment (MauroToscano) [inline batcher/aligned-batcher/src/lib.rs]:
```suggestion
    async fn flush_queue_and_clear_nonce_cache(&self) {
```

Comment (taturosati) [reply batcher/aligned-batcher/src/lib.rs]: changed on 7700d8197cd52aacb695bb885f428f741532665a

**Review by uri-99 — CHANGES_REQUESTED** [review]
If a tx reverts because of invalid nonce, batcher will try to send the same tx to blockchain ~50 times before it stops trying. It should try only once or twice, because no variable is changing between retries.

The following error appeared every times in the batcher when sending concurrent proofs
`ERROR aligned_batcher] Error while sending batch inclusion data response: IO error: Broken pipe (os error 32)`

Batcher gets the following error when trying to send second message to client
`Unexpected error: IO error: Connection reset by peer (os error 54)`

**Review by MauroToscano — APPROVED** [review]

**Review by uri-99 — APPROVED** [review]
I understand my previous comments should be addressed in another PR

**Pull request merged**